Review followup: PR #23882 — For #23854: keep triage prelaunch failures retryable

## Unaddressed review bot suggestions

PR #23882 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/23882

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/pulse-ancillary-dispatch.sh:1043`
- `.agents/scripts/pulse-ancillary-dispatch.sh:355`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/pulse-ancillary-dispatch.sh:355`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/23882#discussion_r3274649747)

```diff
+# Outputs the infrastructure failure reason, or nothing when not recognised.
+# Returns 0 always.
+#######################################
+_triage_runtime_infra_failure_reason() {
+	local sample="$1"
+
+	if printf '%s' "$sample" | grep -qE 'Canary test FAILED|Canary failed.*aborting dispatch' 2>/dev/null; then
+		printf '%s\n' 'canary-unavailable'
+		return 0
+	fi
+
+	if printf '%s' "$sample" | grep -qE 'WORKER_ISSUE_NUMBER unset|WORKER_WORKTREE_PATH unset|issue worker env contract missing|worker --dir does not match WORKER_WORKTREE_PATH|worker worktree repo mismatch' 2>/dev/null; then
```

> ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> The infrastructure failure classification regex is missing several prelaunch failure modes defined in `headless-runtime-helper.sh`. Specifically, `WORKER_WORKTREE_PATH does not exist` (line 338), `opencode version mismatch` (line 1605), and `launch cwd is deleted` (line 289) are not caught. Note: the generic keyword 'issue' has been removed from the search pattern to avoid false positives per project rules.
> 
> ```suggestion
> 	if printf '%s' "$sample" | grep -qE 'WORKER_ISSUE_NUMBER unset|WORKER_WORKTREE_PATH unset|worker env contract missing|worker --dir does not match WORKER_WORKTREE_PATH|worker worktree repo mismatch|WORKER_WORKTREE_PATH does not exist|opencode version mismatch|launch cwd is deleted' 2>/dev/null; then
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Avoid using the word 'issue' as an actionable keyword in search/filter logic as it is too generic and leads to false positives.
> </details>

#### gemini-code-assist on `.agents/scripts/pulse-ancillary-dispatch.sh:1043`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/23882#discussion_r3274649777)

```diff
+	# does not reject triage sessions before model launch (GH#23854).
 	# shellcheck disable=SC2086
 	"$HEADLESS_RUNTIME_HELPER" run \
-		--role worker \
+		--role triage \
 		--session-key "triage-review-${issue_num}" \
 		--dir "$repo_path" \
 		$model_flag \
 		--agent triage-review \
 		--title "Sandboxed triage review: Issue #${issue_num}" \
-		--prompt-file "$prefetch_file" </dev/null >"$review_output_file" 2>&1
+		--prompt-file "$prefetch_file" </dev/null >"$review_output_file" 2>&1 || true
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> To ensure accurate metrics and logging within the headless runtime, the triage worker should export the standard environment variables (`WORKER_ISSUE_NUMBER`, `WORKER_REPO_SLUG`, and `WORKER_WORKTREE_PATH`). Additionally, setting `HEADLESS=1` ensures that the runtime helper enables diagnostic logging. Guard clauses have been added to ensure these variables are not empty before execution.
> 
> ```suggestion
> 	[[ -n "$issue_num" && -n "$repo_slug" && -n "$repo_path" ]] && \
> 	env HEADLESS=1 WORKER_ISSUE_NUMBER="$issue_num" WORKER_REPO_SLUG="$repo_slug" WORKER_WORKTREE_PATH="$repo_path" \
> 	"$HEADLESS_RUNTIME_HELPER" run \
> 		--role triage \
> 		--session-key "triage-review-${issue_num}" \
> 		--dir "$repo_path" \
> 		$model_flag \
> 		--agent triage-review \
> 		--title "Sandboxed triage review: Issue #${issue_num}" \
> 		--prompt-file "$prefetch_file" < /dev/null > "$review_output_file" 2>&1 || true
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Add guard clauses to check if a variable is empty before using it, especially when derived from external sources.
> </details>

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/23882#pullrequestreview-4329313137)

> ## Code Review
> 
> This pull request introduces a distinct 'triage' role for the headless runtime and implements a classification system for infrastructure and contract failures. These changes ensure that infrastructure-related issues do not consume the triage retry budget or incorrectly label GitHub issues as failed. Feedback from the review includes expanding the failure classification regex to cover additional prelaunch failure modes and adding guard clauses with environment variables to the triage worker dispatch for better diagnostic logging and reliability.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.17.18 automated scan.

--- a/.agents/scripts/tests/test-triage-review-worker-contract.sh
+++ b/.agents/scripts/tests/test-triage-review-worker-contract.sh
@@ -27,8 +27,8 @@ fail() {
 write_headless_helper() {
 	cat >"${TEST_TMP}/headless-runtime-helper.sh" <<'EOS'
 #!/usr/bin/env bash
-printf 'WORKER_ISSUE_NUMBER=%s WORKER_REPO_SLUG=%s WORKER_WORKTREE_PATH=%s %s\n' \
-	"${WORKER_ISSUE_NUMBER:-<unset>}" "${WORKER_REPO_SLUG:-<unset>}" "${WORKER_WORKTREE_PATH:-<unset>}" "$*"
+printf 'HEADLESS=%s WORKER_ISSUE_NUMBER=%s WORKER_REPO_SLUG=%s WORKER_WORKTREE_PATH=%s %s\n' \
+	"${HEADLESS:-<unset>}" "${WORKER_ISSUE_NUMBER:-<unset>}" "${WORKER_REPO_SLUG:-<unset>}" "${WORKER_WORKTREE_PATH:-<unset>}" "$*"
 exit 0
 EOS
 	chmod +x "${TEST_TMP}/headless-runtime-helper.sh" || fail "failed to chmod headless helper"
@@ -69,6 +69,9 @@ _run_triage_review_worker "42" "owner/repo" "${TEST_TMP}/repo" "" "$prefetch_fil
 	fail "valid worker launch should not fail the caller"
 if ! grep -q 'WORKER_ISSUE_NUMBER=42' "$valid_output"; then
 	fail "valid worker launch did not invoke the headless runtime"
+fi
+if ! grep -q 'HEADLESS=1' "$valid_output"; then
+	fail "valid worker launch did not export headless diagnostic mode"
 fi
 if ! grep -q -- '--prompt-file' "$valid_output"; then
 	fail "valid worker launch did not pass the prompt file flag"

